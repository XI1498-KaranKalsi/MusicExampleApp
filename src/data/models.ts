export interface Song {
  wrapperType?: 'audiobook' | 'track',
  artistId?: number;
  collectionId?: number;
  artistName: string;
  collectionName?: string;
  collectionCensoredName?: string;
  artistViewUrl?: string;
  collectionViewUrl?: string;
  artworkUrl60: string;
  artworkUrl100: string;
  collectionPrice: number;
  trackCount?: number;
  copyright?: string;
  releaseDate: string;
  primaryGenreName: string;
  previewUrl: string;
  description?: string;
  amgArtistId?: number;
  trackId?: number;
  trackName?: string;
  trackCensoredName?: string;
  trackViewUrl?: string;
  artworkUrl30?: string;
  trackPrice?: number;
  discCount?: number;
  discNumber?: number;
  trackNumber?: number;
  trackTimeMillis?: number;
  contentAdvisoryRating?: string;
  isStreamable?: boolean;
  collectionArtistId?: number;
  collectionArtistName?: string;
  collectionArtistViewUrl?: string;
  trackRentalPrice?: number;
  collectionHdPrice?: number;
  trackHdPrice?: number;
  trackHdRentalPrice?: number;
  longDescription?: string;
  hasITunesExtras?: boolean;
  shortDescription?: string;
}

export interface SearchSongsResponse {
  resultCount: number;
  results: Song[];
}
