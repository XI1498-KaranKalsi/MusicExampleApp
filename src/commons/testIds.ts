const testIds = {
    SongsList: {
        Loader: 'SongList_Loader',
        Flatlist: 'SongList_FlatList'
    },
    SongItem: {
        Title: 'SongItem_Title',
        Artist: 'SongItem_Artist',
        Artwork: 'SongItem_Artwork'
    }
}

export default testIds;