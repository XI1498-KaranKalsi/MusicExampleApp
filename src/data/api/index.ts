import wretch, {ResponseChain, Wretcher} from 'wretch';
import {SearchSongsResponse, Song} from '../models';
import endpoints from './endpoints';

export const BASE_URL = 'https://itunes.apple.com/';

export const apiUrl = (endpoint: string) => `${BASE_URL}${endpoint}`;

const searchSongs = async (term: string): Promise<Song[]> => {
  const res: ResponseChain = await wretch(apiUrl(endpoints.searchSongs))
    .query({
      term,
    })
    .get();
  const data = await res.json<SearchSongsResponse>();
  return data.results.filter(song=>song.wrapperType === 'track');
};

export default {
  searchSongs,
};
