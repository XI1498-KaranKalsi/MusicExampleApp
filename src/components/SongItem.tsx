import React, {FunctionComponent} from 'react';
import {
  Image,
  Text,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
} from 'react-native';
import testIds from '../commons/testIds';
import {Song} from '../data/models';
import SongItemStyles from './styles/SongItemStyles';

export interface SongItemProps extends TouchableOpacityProps {
  song: Song;
  onSongPress?: (song: Song) => void;
}
const SongItem: FunctionComponent<SongItemProps> = props => {
  const {song, onSongPress} = props;
  const onPressCallback = React.useCallback(
    () => onSongPress && onSongPress(song),
    [onSongPress, song],
  );
  return (
    <TouchableOpacity
      style={SongItemStyles.container}
      onPress={onPressCallback}>
      <Image
        testID={testIds.SongItem.Artwork}
        style={SongItemStyles.artwork}
        source={{uri: song.artworkUrl60}}
      />
      <View style={SongItemStyles.infoContainer}>
        <Text
          testID={testIds.SongItem.Title}
          style={SongItemStyles.title}
          numberOfLines={2}>
          {song.trackName}
        </Text>
        <Text
          testID={testIds.SongItem.Artist}
          style={SongItemStyles.artist}>{`Artist   ${song.artistName}`}</Text>
      </View>
    </TouchableOpacity>
  );
};
export default SongItem;
