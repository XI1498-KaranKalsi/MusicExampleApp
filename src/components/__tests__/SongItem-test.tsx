import React from 'react';
import renderer from 'react-test-renderer';
import {cleanup, render, waitFor} from '@testing-library/react-native';
import SongItem, {SongItemProps} from '../SongItem';
import searchSongsStub from '../../stub/searchSongs.json';
import testIds from '../../commons/testIds';
import { Song } from '../../data/models';

describe('testing ui component SongItem', () => {
  const defaultProps: SongItemProps = {
    song: (searchSongsStub.results as Song[])[0],
  };

  beforeEach(() => {});
  afterEach(cleanup);
  it('rendering correctly', () => {
    renderer.create(<SongItem {...defaultProps} />);
  });
  it('matches with snapshot', () => {
    const {toJSON} = render(<SongItem {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });
  it('data loading correctly', () => {
    const {getByTestId} = render(<SongItem {...defaultProps} />);
    expect(getByTestId(testIds.SongItem.Title).props.children).toBe(defaultProps.song.trackName);
    expect(getByTestId(testIds.SongItem.Artist).props.children).toContain(defaultProps.song.artistName);
    expect(getByTestId(testIds.SongItem.Artwork).props.source.uri).toBe(defaultProps.song.artworkUrl60); 
  });
});
