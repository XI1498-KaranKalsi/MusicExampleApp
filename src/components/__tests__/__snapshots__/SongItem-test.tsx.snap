// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ui component SongItem matches with snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderBottomColor": "#d3d3d3",
      "borderBottomWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "opacity": 1,
      "paddingVertical": 12,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://is2-ssl.mzstatic.com/image/thumb/Video/v4/54/84/96/548496ea-06f7-0276-7ab1-6592ca217a29/source/60x60bb.jpg",
      }
    }
    style={
      Object {
        "height": 60,
        "marginHorizontal": 10,
        "width": 60,
      }
    }
    testID="SongItem_Artwork"
  />
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <Text
      numberOfLines={2}
      style={
        Object {
          "color": "#1d7493",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
      testID="SongItem_Title"
    >
      Michael Jackson's This Is It
    </Text>
    <Text
      style={
        Object {
          "color": "#1d7493",
          "fontSize": 14,
          "marginTop": 8,
        }
      }
      testID="SongItem_Artist"
    >
      Artist   Kenny Ortega
    </Text>
  </View>
</View>
`;
